core/embed/rust/src/protobuf/decode.rs

```diff
@@ -289,17 +289,125 @@ impl<'a> InputStream<'a> {
+        for byte_count in 1..=5 {
+            let byte: u32 = self.read_byte()? as u32;
+            let continue_decoding = byte & 0x80 != 0;
+            let payload = byte & 0x7F;
+            if byte_count == 5 {
+                // first 4 bytes are accepted as is (4 x 7 bits = 28 bits),
+                // then the 5th byte is only accepted it it is the final byte
+                // and contains no more than 4 useful bits! (4 x 7 bits + 4 bits = 32 bits)!
+                if continue_decoding || payload > 0xF {
+                    return Err(error::overflow());
+                }
+            }
+            uint += payload << shift;
+            shift += 7;
+            if !continue_decoding {
+                break;
+            }
+        }
+        Ok(uint)
+    }
+
+    pub fn read_uvarint_u64(&mut self) -> Result<u64, Error> {
+        let mut uint: u64 = 0;
+        let mut shift = 0;
+        for byte_count in 1..=10 {
+            let byte: u64 = self.read_byte()? as u64;
+            let continue_decoding = byte & 0x80 != 0;
+            let payload = byte & 0x7F;
+            if byte_count == 10 {
+                if continue_decoding || payload > 1 {
+                    return Err(error::overflow());
+                }
+            }
+            uint += payload << shift;
             shift += 7;
-            if byte & 0x80 == 0 {
+            if !continue_decoding {
                 break;
             }
         }
         Ok(uint)
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn read_uvarint_32() {
+        // 1 byte...
+        let mut stream = InputStream::new(&[0b00000001]);
+        let result = stream.read_uvarint_u32();
+        assert!(result.is_ok());
+        assert_eq!(result.unwrap(), 1);
+
+        // 2 bytes...
+        let mut stream = InputStream::new(&[0b10010110, 0b00000001]);
+        let result = stream.read_uvarint_u32();
+        assert!(result.is_ok());
+        assert_eq!(result.unwrap(), 150);
+
+        // a 0 taking up 5 whole bytes...
+        let mut stream = InputStream::new(&[0b10000000, 0b10000000, 0b10000000, 0b10000000, 0b00000000]);
+        let result = stream.read_uvarint_u32();
+        assert!(result.is_ok());
+        assert_eq!(result.unwrap(), 0);
+
+        // 5 bytes, with only 4 bits used in the last byte... still OK!
+        let mut stream = InputStream::new(&[0b10000000, 0b10000000, 0b10000000, 0b10000000, 0b00001111]);
+        let result = stream.read_uvarint_u32();
+        assert!(result.is_ok());
+        assert_eq!(result.unwrap(), 0xF0000000);
+
+        // 5 bytes with more than 4 bits used in the last byte => overflow
+        let mut stream = InputStream::new(&[0b10000000, 0b10000000, 0b10000000, 0b10000000, 0b00010000]);
+        assert!(stream.read_uvarint_u32().is_err());
+
+        // 5+ bytes => overflow
+        let mut stream = InputStream::new(&[0b10000000, 0b10000000, 0b10000000, 0b10000000, 0b10000000]);
```

Comment:
please make sure that the encoded value is valid -- i.e., if `read_uvarint_u32` is _not_ correctly overflowing here, maybe the error is because the number is too short

maybe add a testcase that reads the same number as u32 and u64 and checks that one succeeds and the other one fails?